File: backend/app/schemas/subscription.py
"""
订阅相关 Schema
"""
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, Field

from app.models.rss import RssCategory


class SubscriptionCreate(BaseModel):
    """创建订阅请求"""
    rss_source_id: int
    is_active: bool = True
    custom_fetch_interval: Optional[int] = Field(None, ge=5, le=1440)


class SubscriptionUpdate(BaseModel):
    """更新订阅请求"""
    is_active: Optional[bool] = None
    custom_fetch_interval: Optional[int] = Field(None, ge=5, le=1440)


class SubscriptionResponse(BaseModel):
    """订阅响应"""
    id: int
    rss_source_id: int
    is_active: bool
    custom_fetch_interval: Optional[int]
    created_at: datetime

    # 关联的 RSS 源信息
    rss_source_name: str
    rss_source_url: str
    rss_source_category: RssCategory
    rss_source_description: Optional[str]
    entry_count: int
    unread_count: int
    last_fetched_at: Optional[datetime]
    last_fetch_status: str

    model_config = {"from_attributes": True}


class SubscriptionListResponse(BaseModel):
    """订阅列表响应"""
    items: List[SubscriptionResponse]
    total: int


class BatchSubscribeRequest(BaseModel):
    """批量订阅请求"""
    rss_source_ids: List[int]


class BatchSubscribeResponse(BaseModel):
    """批量订阅响应"""
    success: List[int]  # 成功订阅的 rss_source_id
    already_subscribed: List[int]  # 已经订阅的
    not_found: List[int]  # 不存在的


class RssMarketItemResponse(BaseModel):
    """RSS 市场条目响应"""
    id: int
    name: str
    url: str
    website_url: Optional[str]
    description: Optional[str]
    category: RssCategory
    icon_url: Optional[str]
    entry_count: int
    # 用户是否已订阅
    is_subscribed: bool

    model_config = {"from_attributes": True}


class RssMarketListResponse(BaseModel):
    """RSS 市场列表响应"""
    items: List[RssMarketItemResponse]
    total: int


class RssBatchImportItem(BaseModel):
    """批量导入条目"""
    name: str = Field(..., min_length=1, max_length=100)
    url: str = Field(..., min_length=1, max_length=500)
    category: RssCategory = RssCategory.OTHER
    description: Optional[str] = None


class RssBatchImportRequest(BaseModel):
    """批量导入请求"""
    items: List[RssBatchImportItem]


class RssBatchImportResponse(BaseModel):
    """批量导入响应"""
    success: int
    failed: int
    errors: List[str]

File: frontend/src/types/subscription.ts
// Backend RSS category enum
export type RssCategory = 'blog' | 'community' | 'paper' | 'social' | 'news_podcast' | 'other';

// Backend subscription response
export interface Subscription {
  id: number;
  rss_source_id: number;
  is_active: boolean;
  custom_refresh_time: string | null;  // 每日刷新时间，如 "08:00"
  created_at: string;
  rss_source_name: string;
  rss_source_url: string;
  rss_source_category: RssCategory;
  rss_source_description: string | null;
  entry_count: number;
  unread_count: number;
  last_fetched_at: string | null;
  last_fetch_status: string;
}

// 可用的刷新时间选项
export const REFRESH_TIME_OPTIONS = [
  { value: 'default', label: 'Default' },
  { value: '06:00', label: '6:00 AM' },
  { value: '08:00', label: '8:00 AM' },
  { value: '09:00', label: '9:00 AM' },
  { value: '12:00', label: '12:00 PM' },
  { value: '18:00', label: '6:00 PM' },
  { value: '20:00', label: '8:00 PM' },
] as const;

export interface SubscriptionListResponse {
  items: Subscription[];
  total: number;
}

// Backend RSS market item
export interface RssMarketItem {
  id: number;
  name: string;
  url: string;
  website_url: string | null;
  description: string | null;
  category: RssCategory;
  icon_url: string | null;
  entry_count: number;
  is_subscribed: boolean;
}

export interface RssMarketListResponse {
  items: RssMarketItem[];
  total: number;
}

// Frontend Feed type (for display, mapped from Subscription/RssMarketItem)
export interface Feed {
  id: string;
  name: string;
  url: string;
  category: string;
  subscribed: boolean;
  description?: string;
  homepage?: string;
  refreshTime?: string;  // 每日刷新时间，如 "08:00" 或 "default"
  // Keep original data for API calls
  _subscription?: Subscription;
  _marketItem?: RssMarketItem;
}

// All category options for filters - use with i18n: t(`categories.${category}`)
export const CATEGORY_OPTIONS: RssCategory[] = ['blog', 'community', 'paper', 'social', 'news_podcast', 'other'];
